Lib/asyncio/tasks.py

```diff
@@ -564,62 +564,106 @@ async def _cancel_and_wait(fut):
+            raise exceptions.TimeoutError
+        return f.result() if resolve else f
 
-        for f in as_completed(fs):
-            result = await f  # The 'await' may raise.
-            # Use result.
 
-    If a timeout is specified, the 'await' will raise
-    TimeoutError when the timeout occurs before all Futures are done.
+def as_completed(fs, *, timeout=None):
+    """Return an async iterator that yields tasks from the given awaitables
+    in the order they finish as they finish.
 
-    Note: The futures 'f' are not necessarily members of fs.
-    """
-    if futures.isfuture(fs) or coroutines.iscoroutine(fs):
-        raise TypeError(f"expect an iterable of futures, not {type(fs).__name__}")
+        async for completed_task in as_completed(fs):
+            result = await completed_task
+            # Use result.
 
-    from .queues import Queue  # Import here to avoid circular import problem.
-    done = Queue()
+    Supplied Tasks and Futures are yielded as-is once they've completed.
+    Coroutines will be scheduled and their implicitly created tasks will be
+    yielded instead.
 
-    loop = events.get_event_loop()
-    todo = {ensure_future(f, loop=loop) for f in set(fs)}
-    timeout_handle = None
+    A TimeoutError is raised by the async for statement if the supplied
+    timeout is reached before the iterator is exhausted.
 
-    def _on_timeout():
-        for f in todo:
-            f.remove_done_callback(_on_completion)
-            done.put_nowait(None)  # Queue a dummy value for _wait_for_one().
-        todo.clear()  # Can't do todo.remove(f) in the loop.
+    For backwards compatibility, the returned object can also be used as a
+    plain iterator that yields new coroutines representing each next awaitable
+    to be completed:
 
-    def _on_completion(f):
-        if not todo:
-            return  # _on_timeout() was here first.
-        todo.remove(f)
-        done.put_nowait(f)
-        if not todo and timeout_handle is not None:
-            timeout_handle.cancel()
+        for next_completed in as_completed(fs):
+            result = await next_completed
+            # Use result.
 
-    async def _wait_for_one():
-        f = await done.get()
-        if f is None:
-            # Dummy value from _on_timeout().
-            raise exceptions.TimeoutError
-        return f.result()  # May raise f.exception().
+    The coroutines yielded by the plain iterator are not the original
+    awaitables passed in and are the source of TimeoutErrors instead of the
+    for loop.
+    """
+    if inspect.isawaitable(fs):
+        raise TypeError(
+            f"expects an iterable of awaitables, not {type(fs).__name__}"
+        )
 
-    for f in todo:
-        f.add_done_callback(_on_completion)
-    if todo and timeout is not None:
-        timeout_handle = loop.call_later(timeout, _on_timeout)
-    for _ in range(len(todo)):
-        yield _wait_for_one()
+    return _AsCompletedIterator(fs, timeout)
```

Comment:
Perhaps there was a problem when `_AsCompletedIterator` had a `__del__` method? But I removed it because it was not helpful.

Making `as_completed()` a class exposes an implementation detail. In particularly it will allow to use `isinstance()`. I am not sure that it is the final implementation and that we will not have reasons to make it a non-class in future. It will also expose a lot of unneeded details in `help()`.